### PR TITLE
Port mini-swe-agent

### LINKED ISSUE #18: Inspect-native port of SWE Agent
Discussion about ASET contributing an Inspect implementation of SWE Agent can be found in [this thread in Inspect Community Slack](https://inspectcommunity.slack.com/archives/C090U74L4KS/p1759146264247539).

@anaoaktree is currently working on this project - I'm raising this issue for visibility and tracking. Ana's work-in-progress development plan doc is [here](https://docs.google.com/document/d/1uUmXxopQ72TWprBT072GoQnSYFAiSr7xkHlA9jKP4JM/edit?pli=1&tab=t.dgxwmxi9k55c).

### PR DESCRIPTION
**Pull request by anaoaktree**
## Summary

Add mini-swe-agent https://github.com/SWE-agent/mini-swe-agent as a third agent integration for inspect-swe github.com/meridianlabs-ai/inspect_swe, alongside claude_code and codex_cli. Note that mini-swe-agent is python based which required different considerations to existing binary executable agents.

Closes #18 

## [UPDATE] 
- Multi-turn: Subclassed DefaultAgent with a ResumableAgent that loads and parses message history from the trajectory file to restore conversation state across turns. The ResumableAgent is written to the sandbox and loaded via v2's --agent-class flag. Note: each user message includes a clear termination instruction, otherwise the agent may hang.
- Upgraded to support mini-swe-agent v2, removed support for v1. https://mini-swe-agent.com/latest/advanced/v2_migration/
- Migrated to new `exec_remote` from inspect

##  Results over one run for each task
Compared with [terminal bench reported results using gpt-5-mini](https://www.tbench.ai/leaderboard/terminal-bench/2.0/mini-swe-agent/unknown/gpt-5-mini%40openai
)
### Random five challenges
```
uv run --no-sync inspect eval examples/terminal_bench_test --limit 5 --model openai/gpt-5-mini
```
Screenshot 2026-01-30 at 13 22 46

Results in line with TB 

### Four pass challenges
```
uv run --no-sync inspect eval examples/terminal_bench_test --model openai/gpt-5-mini  -T eval_names='["constraints-scheduling", "fix-git", "git-multibranch", "prove-plus-comm"]'
```
Screenshot 2026-01-30 at 13 20 54

Results in line with TB except for `fix-git`.

### mini-swe-agent CLI
- New `_mini_swe_agent/mini_swe_agent.py` file utilising most of the same control structures as existing agents. 
- Currently pinned to `1.17.4` which is the latest v1 stable version, monitoring v2.0 release.
- Consider adding version validation to warn if user requests v2.x (once released) 

```bash
mini \
  --yolo \              # Run without confirmations
  --exit-immediately \  # Exit when done (non-interactive)
  --cost-limit N \      # Optional cost limit
  --task "PROMPT"       # The task to execute
```

### Package Installation Pattern
**pip wheel download on host → tarball → uv offline install in sandbox** 

- using pip to download wheels on host, uv to install for sandboxes without python.
- Follows the similar host-download/sandbox-install pattern as binary agents. Mirrored some of the existing utilities but for wheels 
- Detects platform and python version, pip manages version resolution

**Note**: Created new `_util/agentwheel.py` file with potential to expand into a more robust shared utility for other python agents with improved data structures. Suggested improvement after 2nd python agents gets ported.

## Tests
- Added `test_download.py` and `test_ensure_wheels_install.py` to test sandbox and package setup (no need for API keys)
- Extended `test_system_explorer.py` to support mini-swe-agent
- Created `examples/terminal_bench_test/` to test terminal bench tasks with the agents for validation. Challenges were picked based on their consistency on the official leaderboard and cost running. Picked [break-filter-js-from-html](https://www.tbench.ai/leaderboard/terminal-bench/2.0/mini-swe-agent/unknown/gpt-5-mini%40openai/b33421492e86b07de2a01bc2205f7e7d2fdc36446a84787265d99e5e138fdaaa) that always failed and [constraints-scheduling](https://www.tbench.ai/leaderboard/terminal-bench/2.0/mini-swe-agent/unknown/gpt-5-mini%40openai/c69027b9099bb7ac63a753b310e31cfef9cc20dce2d1c6c88956c8a854d7ac16) that always passed.

### No built-in session
- mini-swe-agent doesn't natively handle multi-turn conversation ([From their docs](https://mini-swe-agent.com/latest/faq/): _There is no running shell session. Every action is executed as a subprocess, that means every action is independent of the previous ones (it is literally a subprocess.run/os.system/docker exec call)._ )
- [WIP] Current version handles single turn prompts (--task option). To be modified for use with multiple attempts. Options include save trajectory file to --output constructing history manually. Requires further investigation

### Testing
- Only tested with gpt-5-mini - Support/test local models
- Extend new tests for existing agents `claude_code` and `codex_cli` and improve code reusability

## Next steps & Open questions
- Feedback from this PR
- Extend testing - Could add more tests from existing evals to validate agent outcome (not just agent infrastructure) and expand to existing agents
- Working on the limitations
- Potential to simplify common control structures (needs more investigation, considered out of scope for now) and simplify some of the longer functions.

### CODE REVIEW

File: src/inspect_swe/_mini_swe_agent/mini_swe_agent.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                "--yolo",  # run without confirmations (like --print for claude)
+                "--exit-immediately",  # exit when agent finishes instead of prompting
+            ]
+
+            # add cost limit if specified
+            if cost_limit is not None:
+                cmd.extend(["--cost-limit", str(cost_limit)])
+
+            # build user prompt
+            prompt, has_assistant_response = build_user_prompt(state.messages)
```
Comment (celiawaggoner) [inline src/inspect_swe/_mini_swe_agent/mini_swe_agent.py]: is `has_assistant_response` ever used / needed?

File: src/inspect_swe/_mini_swe_agent/mini_swe_agent.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                system_messages.append(system_prompt)
+            if system_messages:
+                # Prepend system context to the task
+                system_context = "\n\n".join(system_messages)
+                full_prompt = (
+                    f"System instructions:\n{system_context}\n\nTask:\n{prompt}"
+                )
+
+            # resolve sandbox
+            sbox = sandbox_env(sandbox)
```
Comment (celiawaggoner) [inline src/inspect_swe/_mini_swe_agent/mini_swe_agent.py]: `sandbox_env(sandbox)` is called twice (here and line 109). seems like a minor inefficiency - is it possible to refactor to just call it once?

File: src/inspect_swe/_mini_swe_agent/mini_swe_agent.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+from inspect_ai.util import store
+
+from .._util._async import is_callable_coroutine
+from .._util.agentwheel import AgentWheelSource, ensure_agent_wheel_installed
+from .._util.messages import build_user_prompt
+from .._util.trace import trace
+
+# mini-swe-agent wheel source configuration
+# Pin to v1.x by default - v2.0 has breaking changes (migration guide pending)
+# See: https://mini-swe-agent.com/latest/advanced/v2_migration/
```
Comment (celiawaggoner) [inline src/inspect_swe/_mini_swe_agent/mini_swe_agent.py]: 👍 thanks for documenting!

File: tests/test_download.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+@pytest.mark.parametrize(
+    "version,platform",
+    [
+        ("1.16.0", "linux-x64"),
+        (None, "linux-x64"),  # Latest
+        ("1.17.4", "linux-arm64"),
+        ("1.17.4", "linux-x64"),
+    ],
+)
+def test_mini_swe_agent_wheels_download(
```
Comment (celiawaggoner) [inline tests/test_download.py]: also clean up the cache once the tests are run?

Comment (celiawaggoner) [reply tests/test_download.py]: and some edge cases to consider adding tests for:
- Invalid version strings
- Unsupported platforms
- Network failures (with mocking)
- Cache hit path (download twice, verify second is from cache)

File: tests/test_ensure_wheels_install.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+from inspect_ai.solver import Generate, Solver, TaskState, solver
+from inspect_ai.util import sandbox
+from inspect_swe._mini_swe_agent.mini_swe_agent import MINI_SWE_AGENT_SOURCE
+from inspect_swe._util.agentwheel import ensure_agent_wheel_installed
+
+from tests.conftest import skip_if_no_docker
+
+
+@solver
+def install_mini_swe_agent(version: str = "1.17.4") -> Solver:
```
Comment (celiawaggoner) [inline tests/test_ensure_wheels_install.py]: nit: `version` is hard-coded in two places, so you could extract it out into a constant

File: tests/test_ensure_wheels_install.py
Comment (celiawaggoner) [inline tests/test_ensure_wheels_install.py]: add test cases for `version="stable"` and `version="latest"`?

Comment (celiawaggoner) [reply tests/test_ensure_wheels_install.py]: and a test case to confirm an invalid version raises the appropriate error?

File: src/inspect_swe/_mini_swe_agent/mini_swe_agent.py (hunk 1 above)
Comment (anaoaktree) [reply src/inspect_swe/_mini_swe_agent/mini_swe_agent.py]: @celiawaggoner not in this case (neither is the while True structure) because of the limitation of the single-turn conversation (in the PR description). I left it mostly to keep the structure of the other agents and to then extend to the multi-turn version should you think that is part of the scope for this project. Otherwise I'll clean this function up a bit.

File: src/inspect_swe/_mini_swe_agent/mini_swe_agent.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+                result = await sbox.exec(
+                    cmd=["bash", "-c", 'exec 0</dev/null "$@"', "bash"] + agent_cmd,
+                    cwd=cwd,
+                    env={
+                        "MSWEA_CONFIGURED": "true",  # Skip interactive setup wizard
+                        "MSWEA_MODEL_NAME": model
+                        if model is not None
+                        else get_model().name,
+                        "OPENAI_API_BASE": f"http://localhost:{bridge.port}/v1",
+                        "OPENAI_API_KEY": "sk-none",
```
[the remark this answers is not in the document]
Comment (anaoaktree) [reply src/inspect_swe/_mini_swe_agent/mini_swe_agent.py]: it is just so the agent detects a key, the real one will be passed by inspect.

File: examples/terminal_bench_test/task.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+    raise ImportError(
+        "Could not find Terminal Bench 2.0 challenges directory. "
+        "Please install inspect-evals[terminal_bench_2]:\n"
+        "  pip install inspect-evals[terminal_bench_2]"
+    )
+
+
+@task
+def terminal_bench_task(
```
[the remark this answers is not in the document]
Comment (Jay-Bailey) [reply examples/terminal_bench_test/task.py]: I can confirm this command now runs, even without the --T.

**Review by Jay-Bailey — APPROVED** [review]
This now looks good and the command runs. I am not sure what my approval does to your ability to merge, but I trust that Meridian has a setup they're happy with around it. I am happy with the PR. Thanks for your patience!

**Comment by jjallaire** [comment]
Okay, no worries, just ping me when this is ready for final review. Thanks!
